### PR TITLE
[RFC] find_term_bykeys to return bool instead of ssize_t

### PR DESCRIPTION
**Pull request by equalsraf**
- Return value is true if found, false otherwise
- index value (size_t) is set using ptr argument
- from #1749 - to avoid the use of ssize_t

@elmart I ended up going for returning a bool. It seemed more idiomatic based on the places where the function is being used. But the compound return case seems fine too, and if we can get some consensus on naming I don't mind getting the ball rolling in this PR.

### CODE REVIEW

**Comment by elmart** [comment]
Ok. Maybe better to leave the compound return thing for a PR on its own, if there 's consensus on that. 
LGTM.

**Comment by aktau** [comment]
In my opinion, both the custom compound returns and the out-pointer argument are a bit icky. I think I'd prefer conditionally defining `ssize_t` for Windows so that the current version keeps working.

About the creation of compound values with macro's: I've wanted to speak up on the continued macroization of the codebase and the loss of readability that results from that. Macro's definitely have their uses and place, they can be very handy and sometimes do otherwise impossible (in C) things (see klist/khash/...), but I think we may be overusing it a bit.

That said, of the two, the pointer out parameter I find to be the more palatable of the two. 

Note that using `ssize_t` to represent a value + error (algebraic sum datatype hack) is something used by POSIX as well: http://man7.org/linux/man-pages/man2/read.2.html

**Comment by elmart** [comment]
Well, IMO, macros, if used right, do increase readability, not the other way around.
I know this is a very subjective thing, though. What someone considers 'readable' depends much on background, habits, etc. That's why I said 'if consensus is reached'.

**Comment by equalsraf** [comment]
> In my opinion, both the custom compound returns and the out-pointer argument are a bit icky. I think I'd prefer conditionally defining ssize_t for Windows so that the current version keeps working.

Ups I just realized I mistyped in the other thread - ssize_t is defined for windows, SSIZE_MAX is what is missing. As far as working for Windows goes I'm using INT_MAX as a replacement right now, but this doesn't seem to be exactly correct because [ssize_t](https://msdn.microsoft.com/en-us/library/windows/desktop/aa383751%28v=vs.85%29.aspx#SSIZE_T) can be either long or int (32/64bits), but both are 4 bytes long anyway. In any case this PR is not a requirement for Windows, just an RFC for clarification if we should invest on this front.

Sidenote: IIIRC **read()** in MSVC uses int both as argument and return value.

> About the creation of compound values with macro's: I've wanted to speak up on the continued macroization of the codebase and the loss of readability that results from that. Macro's definitely have their uses and place, they can be very handy and sometimes do otherwise impossible (in C) things (see klist/khash/...), but I think we may be overusing it a bit.

I agree here - but I don't think we need macros or code generation for compound return values either. But this depends on how many types would need it.

> Well, IMO, macros, if used right, do increase readability, not the other way around.
I know this is a very subjective thing, though. What someone considers 'readable' depends much on background, habits, etc. That's why I said 'if consensus is reached'.

They do increase readability, as in understanding what it does, but decrease it in term of understanding how it does it.

**Comment by justinmk** [comment]
The changes here LGTM.

Somewhat off-topic, are we still resolved to define `ssize` as `intptr_t` as discussed in https://github.com/neovim/neovim/issues/682#issuecomment-44571076 ?

**Comment by elmart** [comment]
@justinmk I think `ptrdiff_t` was the chosen one there, not `intptr_t`.
In any case, I think we could do better now:
- Use CMAKE to define missing SSIZE_MAX, and use ssize_t.
- Better yet: If possible, like done here, get rid of ssize_t altogether. There are very few uses of it.

File: src/nvim/term.c [hunk 1]
```diff
@@ -3825,11 +3825,11 @@ ssize_t find_term_bykeys(char_u *src)
     assert(termcodes[i].len >= 0);
     if (slen == (size_t)termcodes[i].len
         && STRNCMP(termcodes[i].code, src, slen) == 0) {
-      assert(i <= SSIZE_MAX);
-      return (ssize_t)i;
+      *idx = i;
```
Comment (otommod) [inline src/nvim/term.c]: Wouldn't it be good to check for `NULL` before assigning?

**Comment by equalsraf** [comment]
@otommod thanks, added a check for NULL

**Comment by oni-link** [comment]
Can this be closed? `term.c` is gone.

**Pull request closed without merging**